state/protocol/protocol_state/kvstore/models.go

```diff
@@ -34,14 +45,33 @@ type modelv0 struct {
 	upgradableModel
 }
 
-var _ protocol_state.Reader = new(modelv0)
-var _ protocol_state.API = new(modelv0)
+var _ protocol_state.KVStoreAPI = (*modelv0)(nil)
+var _ protocol_state.KVStoreMutator = (*modelv0)(nil)
 
+// ID returns an identifier for this key-value store snapshot by hashing internal fields.
 func (model *modelv0) ID() flow.Identifier { return flow.MakeID(model) }
 
-func (model *modelv0) Clone() protocol_state.API {
-	cpy := *model
-	return &cpy
+// Replicate instantiates a Protocol State Snapshot of the given `protocolVersion`.
+// It clones existing snapshot if `protocolVersion = 0` and performs a migration if `protocolVersion = 1`.
+// Expected errors during normal operations:
+//   - ErrIncompatibleVersionChange if replicating the Parent Snapshot into a Snapshot
+//     with the specified `protocolVersion` is not supported.
+func (model *modelv0) Replicate(protocolVersion uint64) (protocol_state.KVStoreMutator, error) {
```

Comment:
We discussed having `Replicate` live alongside the models and be an "atomic" process, in the sense that we will strictly return a valid KVStore model from `Replicate`. 

In my mind that constraint includes making sure that a KVStore which has been upgraded to protocol version N does not include a pending `VersionUpgrade` for protocol version N. However we're not doing that here.

I'm wondering if `Replicate` should instead receive as input the **block view**, and internally decide whether an upgrade is needed, based on the `VersionUpgrade` field? 

That way, the higher-level logic would not need to deal with:
- deciding when the version becomes active
- removing the old `VersionUpgrade` following the upgrade

Based on [your comment here](https://github.com/onflow/flow-go/pull/5513/files/856299a96640f57110e2a792cf59bc9bcc9a6f9d..ba1f763f08d56fff590b4c9c3e70399c72749f45#diff-d5d9a8e1e06fcaf771c56b400d81c56429405675f36121b42294c3cb3eee38c0R90), I assume your plan is to keep outdated `VersionUpgrade`s in the protocol state until they are replaced. If we choose that route, this comment is mostly irrelevant (we could still choose to pass in view rather than `protocolVersion` to `Replicate`).

Overall, I'm happy to merge this as-is and make a final decision when we hook up the state machine.

Reply:
I was thinking about cleaning `VersionUpgrade` when we replicate, but the process still waits for trigger. Happy to discuss after. For now leaving as it is.